apps/recnet-api/src/modules/article/article.service.ts

```diff
@@ -19,39 +20,68 @@ export class ArticleService {
     private readonly digitalLibraryService: DigitalLibraryService | null
   ) {}
 
+  // public async getArticleByLink(
+  //   link: string
+  // ): Promise<GetArticleByLinkResponse> {
+  //   let unifiedLink = link;
+  //
+  //   // If digital library service is available, try to get the unified link
+  //   if (this.digitalLibraryService) {
+  //     unifiedLink = await this.digitalLibraryService.getUnifiedLink(link);
+  //   }
+  //
+  //   // Try to find the article in the database using the unified link
+  //   let article = await this.articleRepository.findArticleByLink(unifiedLink);
+  //   if (article) {
+  //     return { article };
+  //   }
+  //
+  //   // If no article found, try to get metadata using the digital library service
+  //   if (this.digitalLibraryService) {
+  //     try {
+  //       const metadata = await this.digitalLibraryService.getMetadata(link);
+  //       const articleInput = this.transformMetadata(metadata, unifiedLink);
+  //
+  //       article = await this.articleRepository.createArticle(articleInput);
+  //       return { article };
+  //     } catch (error) {
+  //       this.logger.error(error);
+  //     }
+  //   }
+  //
+  //   // If no article found and no metadata available, return null
+  //   return {
+  //     article: null,
+  //   };
+  // }
+
   public async getArticleByLink(
-    link: string
+    link: string,
+    fetchFromDigitalLibrary: boolean = true
```

Comment:
```suggestion
    useDigitalLibraryFallback: boolean = true
```
I think the naming can be better. It's confusing to name like this since if you set `fetchFromDigitalLibrary` to true, it will first fetch from our DB and if not found, it will fetch from Digital Library. Therefore, I think naming `fetchFromDigitalLibrary` is not accurate.

Just provide one of my random idea, we can name it `useDigitalLibraryFallback`. @joannechen1223 any thought?

Reply:
Actually like what I commented out, at first I implemented a separated API to get article directly from the database, but our professor suggests me to combine the getDbArticleByLink with the original getArticleByLink because these two function are pretty similar and he doesn't want the API to explode at the beginning. Also make sense. So I changed the getArticleByLink function to the current version. May need to discuss.

Reply:
I think using separate method might be more scalable in this case given the original method was already really complicated. It also makes sense to try not to make too many APIs but I would prefer make functions easy to maintain than try to add complex logic in a function which is already heavy.
@joannechen1223 thoughts?

Reply:
@yoavartzi Yeah that's the small tradeoff. But I am worried about the over complicated logic inside single function, that's also not easy for future maintainers to understand. IMO, the small amount of duplication is okay.

Reply:
Is there an option to create two APIs but with more code re-use? So the logic is not duplicated. You could think of the two APIs calling the same function that actually implements the logic (and that function itself not being exposed through an API)

Reply:
Sometimes there's this option but sometimes no. I think in this case, it is not really meaningful to do this since, if we are gonna create another API, the function body would look like this, it's already pretty small and not much thing you can abstract away 🤣 . I am actually good with current implementation but please keep/add the comments.

```ts
  let article = await this.articleRepository.findArticleByLink(link);
  if (article) {
    return { article };
  }
  return { article: null };
 ```